### PR TITLE
Cucumber BDD implementation

### PR DESCRIPTION
**Pull request by OlegChub**
What has been done:
- added bdd to the project
- simple test and ddt tests were created
- tests execution parallelization implemented

### CODE REVIEW

**Comment by OlegChub** [comment]
Allure report 
[allure-report.zip](https://github.com/OlegChub/reportPortal/files/9859371/allure-report.zip)

File: src/test/java/cucumber/runner/RunCucumberTest.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package cucumber.runner;
+
+import org.junit.platform.suite.api.ConfigurationParameter;
+import org.junit.platform.suite.api.IncludeEngines;
+import org.junit.platform.suite.api.SelectClasspathResource;
+import org.junit.platform.suite.api.Suite;
+
+import static io.cucumber.core.options.Constants.GLUE_PROPERTY_NAME;
+
+
```
Comment (Fentezi3) [inline src/test/java/cucumber/runner/RunCucumberTest.java]: empty line

File: src/test/java/cucumber/steps/Hooks.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import exeptions.FailedToLoginException;
+import helpers.UserTokenGenerator;
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.BeforeAll;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Hooks {
+    private static final Logger logger = LogManager.getLogger();
```
Comment (Fentezi3) [inline src/test/java/cucumber/steps/Hooks.java]: Maybe the name of the constant should be in upper case?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+public class StepDefinitions {
+    private LaunchController launch = new LaunchController(Hooks.api);
+    private DashboardController dashboards = new DashboardController(Hooks.api);
+    private ArrayList<String> listWithDashboardNames;
+
+    @When("use is logged in")
```
Comment (Fentezi3) [inline src/test/java/cucumber/steps/StepDefinitions.java]: maybe 'User'?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public void useIsLoggedIn() {
```
Comment (Fentezi3) [inline src/test/java/cucumber/steps/StepDefinitions.java]: I'm not sure but maybe better methods with asserts will names like 'checkSomethingImportant'?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        assertNotNull(getToken(), "Authorization failed");
+    }
+
+    @When("user makes request to get all permitted dashboard resources for the project")
+    public void getAllPermittedDashboardResourcesForTheProject() {
+        listWithDashboardNames = dashboards.getAllDashboardNames();
+    }
+
+    @When("user sees dashboard with {string} name")
+    public void getDashboardNameFromList(String dashboardName) {
```
Comment (Fentezi3) [inline src/test/java/cucumber/steps/StepDefinitions.java]: I'm not sure but maybe better methods with asserts will names like 'checkSomethingImportant'?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        assertThat(listWithDashboardNames, hasItem(dashboardName));
+    }
+
+    @When("user makes request to get info of launch with id {int}")
+    public void userMakesGetRequestToGetInfoOfLaunchWithId(int launchId) {
+        launch.getExactLaunchInfo(launchId);
+    }
+
+    @When("user sees that parameter {string} has value {int}")
+    public void userSeesThatParameterHasValue(String parameterName, int expectedValue) {
```
Comment (Fentezi3) [inline src/test/java/cucumber/steps/StepDefinitions.java]: I'm not sure but maybe better methods with asserts will names like 'checkSomethingImportant'?

File: src/test/resources/features/launchInfo.feature [hunk 7]
```diff
@@ -0,0 +1,19 @@
+@api
+Feature: Exact launch info checking
+
+  Background: User is Logged In
+    Given use is logged in
+
+  Scenario: Get dashboard name from the project
```
Comment (Fentezi3) [inline src/test/resources/features/launchInfo.feature]: This scenario is about Dashboards. Maybe better remove them to a separate feature file?

**Review by Fentezi3 — CHANGES_REQUESTED** [review]

File: src/test/java/cucumber/runner/RunCucumberTest.java (hunk 1 above)
Comment (OlegChub) [reply src/test/java/cucumber/runner/RunCucumberTest.java]: removed

File: src/test/java/cucumber/steps/Hooks.java (hunk 2 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java]: done

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 3 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: Sure, typo
Corrected

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 4 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: done

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 5 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: done

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 6 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: done

File: src/test/resources/features/launchInfo.feature (hunk 7 above)
Comment (OlegChub) [reply src/test/resources/features/launchInfo.feature]: done

**Review by Fentezi3 — APPROVED** [review]

File: src/test/java/cucumber/steps/Hooks.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import helpers.UserTokenGenerator;
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.BeforeAll;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Hooks {
+    private static final Logger LOGGER = LogManager.getLogger();
+    public static ApiClient api = new ApiClient();
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/Hooks.java]: fields should be ordered by decreasing their access modifiers

Comment (dKatechev) [reply src/test/java/cucumber/steps/Hooks.java]: why public?

File: src/test/java/cucumber/steps/Hooks.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    private static UserTokenGenerator userToken = new UserTokenGenerator(api);
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/Hooks.java]: Such a strange construction.

File: src/test/java/cucumber/steps/Hooks.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    @BeforeAll
+    public static void setAccessToken() throws FailedToLoginException {
+        LOGGER.info("Setting token ...");
+        userToken.setToken();
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/Hooks.java]: not appropriate name for generator

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.ArrayList;
+
+import static helpers.UserTokenGenerator.getToken;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasItem;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class StepDefinitions {
+    private LaunchController launch = new LaunchController(Hooks.api);
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/StepDefinitions.java]: Why access ApiClient directrly?

File: src/test/resources/features/launchInfo.feature [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    Given user is logged in
+
+  Scenario Outline: Check exact launch defects info
+    When user makes request to get info of launch with id 3
+    Then user sees that parameter "<infoBlockName>" has value <expectedValue>
+    Examples:
+      | infoBlockName | expectedValue |
+      | productBug    | 4             |
+      | toInvestigate | 10            |
+      | automationBug | 4             |
```
Comment (dKatechev) [inline src/test/resources/features/launchInfo.feature]: Required:
Simple test with precondition (e.g. adding a new widget) using test steps with different input parameter types - list, custom object, data table.
Did you implement it?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    public void checkDashboardNameIsInList(String dashboardName) {
+        assertThat(listWithDashboardNames, hasItem(dashboardName));
+    }
+
+    @When("user makes request to get info of launch with id {int}")
+    public void userMakesGetRequestToGetInfoOfLaunchWithId(int launchId) {
+        launch.getExactLaunchInfo(launchId);
+    }
+
+    @When("user sees that parameter {string} has value {int}")
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/StepDefinitions.java]: Maybe use ```@Then``` for verifications?

**Review by dKatechev — CHANGES_REQUESTED** [review]

**Comment by dKatechev** [comment]
Please, review @Fentezi3 PR: https://github.com/Fentezi3/filters-reportportal-framework/pull/4 .

File: src/test/java/cucumber/steps/Hooks.java (hunk 8 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java:14]: change to default

File: src/test/java/cucumber/steps/Hooks.java (hunk 9 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java:15]: why?

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 13 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: But for what? It is steps file, we may use only the one annotation to reduce the import. Another annotations are for feature files

File: src/test/resources/features/launchInfo.feature (hunk 12 above)
Comment (OlegChub) [reply src/test/resources/features/launchInfo.feature]: I thought that user login in the background like a precondition is enough.

File: src/test/java/cucumber/steps/Hooks.java (hunk 9 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java]: I've separated ApiClient for Hooks and for Steps. I think that now it is better

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 11 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: I've separated ApiClient for Hooks and for Steps. I think that now it is better

File: src/test/java/cucumber/steps/Hooks.java (hunk 8 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java]: I've separated ApiClient for Hooks and for Steps. Now it is private

File: src/test/resources/features/launchInfo.feature (hunk 12 above)
Comment (OlegChub) [reply src/test/resources/features/launchInfo.feature]: Do I need to create more tests?

File: src/test/java/cucumber/steps/Hooks.java (hunk 10 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java]: changed classname to UserLogin and method to login()

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 13 above)
Comment (dkatsechau-dano) [reply src/test/java/cucumber/steps/StepDefinitions.java:42]: For better clarity.
[Then steps are used to describe an expected outcome, or result.](https://cucumber.io/docs/gherkin/reference/)

And why should we reduce imports? It doesn't reduce build dependencies, it only clarifies purpose of the step using appropriate annotation.

File: src/test/resources/features/launchInfo.feature (hunk 12 above)
Comment (dkatsechau-dano) [reply src/test/resources/features/launchInfo.feature:14]: Please, add the functionality, described in the Home task:
image

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 13 above)
Comment (dKatechev) [reply src/test/java/cucumber/steps/StepDefinitions.java]: For better clarity.
[Then steps are used to describe an expected outcome, or result.](https://cucumber.io/docs/gherkin/reference/)

And why should we reduce imports? It doesn't reduce build dependencies, it only clarifies purpose of the step using appropriate annotation.

File: src/test/resources/features/launchInfo.feature (hunk 12 above)
Comment (dKatechev) [reply src/test/resources/features/launchInfo.feature]: Please, add the functionality, described in the Home task:
image

File: src/test/java/cucumber/steps/Hooks.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import helpers.UserLogin;
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.BeforeAll;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Hooks {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static ApiClient api = new ApiClient();
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/Hooks.java:14]: You have constuctor with ApiClient argument in many classes, and send there newly created instance of ApiClient each time.
In this case, what is the use to keep it in the constructor?

File: src/test/java/cucumber/steps/StepDefinitions.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+public class StepDefinitions {
+    private ApiClient api = new ApiClient();
+    private LaunchController launch = new LaunchController(api);
+    private DashboardController dashboards = new DashboardController(api);
+    private ArrayList<String> listWithDashboardNames;
+
+    @When("user is logged in")
+    public void checkUserIsLoggedIn() {
+        assertNotNull(getToken(), "Authorization failed");
```
Comment (dKatechev) [inline src/test/java/cucumber/steps/StepDefinitions.java:38]: Whould you add some logging to each step to the log-file readers?

**Review by dKatechev — CHANGES_REQUESTED** [review]
Whould you please add some report to understand, that your tests go smoothly in parallel?

File: src/test/resources/features/launchInfo.feature (hunk 12 above)
Comment (OlegChub) [reply src/test/resources/features/launchInfo.feature]: add test with DataTable

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 13 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: According to my experience in writing test with Cucumber, it worth to pay attention to annotations in features only. Cause feature-files are for bussiness initially. Steps files are for developers, and annotation is not so necessary. But maybe I had bad practice :)

Comment (dKatechev) [reply src/test/java/cucumber/steps/StepDefinitions.java]: You are developer in test automation, please keep layers coherent.

**Comment by OlegChub** [comment]
Add allure report
[allure-report.zip](https://github.com/OlegChub/reportPortal/files/10003247/allure-report.zip)

File: src/test/java/cucumber/steps/Hooks.java (hunk 14 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/Hooks.java:14]: Don't understand the question. In Hooks ApiClient is for login only.

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 13 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java]: update annotations

File: src/test/java/cucumber/steps/StepDefinitions.java (hunk 15 above)
Comment (OlegChub) [reply src/test/java/cucumber/steps/StepDefinitions.java:38]: add logger to checkUserIsLoggedIn method

**Review by dKatechev — APPROVED** [review]

**Pull request merged**